### ISSUE TITLE
Question: Just using EmailBody: is it possible to set the encoding to UTF8?

### ISSUE THREAD

**Issue by schittli** [EvotecIT/PSWriteHTML#123]
Good morning
**Thank you very much for sharing your useful PowerShell Tools!,** it makes playing with PowerShell **much** more fun.

It's also very interesting how you use ScriptBlocks, it makes the usage of your tools very clear :-), but it's hard to understand your nested / abstract PowerShell scripts.

#### I have a question to EmailBody: Can it encode the result as UTF8?
I probably misuse EmailBody, but it would be an elegant solution to just create HTML. I use it like that:
```
$Content = EmailBody -FontFamily 'Calibri' -Size 15 {
   EmailTextBox -FontFamily 'Calibri' -Size 17 {
      'Statusreport'
   }
   EmailText -LineBreak
   New-HTML -Encoding UTF8 {
   New-HTMLTable -DataTable $TblState -DisableInfo `
      -HideFooter -DisablePaging -ScrollCollapse `
      -Buttons @() -DisableSearch -DisableResponsiveTable `
   -DisableStateSave -DisableProcessing
   }
   EmailText -LineBreak
}
```
If I send $Content to a REST API, then I get the error: `Your payload must be encoded in UTF-8.`
Is there a way to make EmailBody returns an UTF8 encoded string?

Thanks a lot for any help, kind regards,
Thomas

**Comment by PrzemyslawKlys** [comment]
Hi,

1. Thank you
2. I'm a bit confused about what is hard to understand and very clear at the same time? :-) Can you elaborate? Maybe I can make it simpler?
3. I actually was planning to allow skipping of sending email to just create email content. It's on my TODO list. It would make it possible to use HTML for other stuff.

I am a bit confused by your example. You are using New-HTML inside EmailBody. Why not simply use New-HTML { ... } and skip EmailBody altogether? Also, I think the encoding is used only in save process. As long as you play with "strings" returned by New-HTML/EmailBody it has no encoding. So most likely you just need to tell REST API to encode it? If I missed something please explain, show examples and so on.

**Comment by schittli** [comment]
Good evening
Thank you very much for your answer!

- The confusion lies here:
For a user, your interface to your library is very nice.
But if one tries to understand your library code, then it's a lot of work because there are many nestings and mappings. There are many things I have not yet understood, but I guess that the code-structure is very well extendable :-)
- _Why not simply use New-HTML { ... } and skip EmailBody altogether?_
I tried to use EmailBody because the HTML generation with EmailBody looked much more elegant in the examples than with New-HTML { ... }.
But as soon as I have used `New-HTMLTable`:

```
$Body = EmailBody -FontFamily 'Calibri' -Size 15 {
   EmailTextBox -FontFamily 'Calibri' {
      'Statusbericht'
   }
   EmailText -LineBreak
   New-HTMLTable -DataTable $TblProcs
}
```
Then I've got: `WARNING: New-HTMLTable - Creation of HTML aborted. Most likely New-HTML is missing.`

Therefore I've embedded `New-HTMLTable { … }` into `New-HTML { … }`

I hope my attempts makes some sense :-)

Kind regards,
Thomas

**Comment by PrzemyslawKlys** [comment]
Ah, ok.

1. Yes, the code is indeed hard to follow but I got used to it now. Unfortunately,, some of my design decisions require a lot of mappings.
2. And this is where the mappings come in. You need New-HTML or Email to deliver your the ability to map things. I will be fixing Email for the ability you describe however I expect this to be the case:

```powershell
Email {
    EmailBody {

    }
}
```
I am not sure I can just do EmailBody, maybe. I will try to. You could always use Dashimo style in the meantime? 

```powershell
Dashboard -Name 'Dashimo Test' -FilePath $PSScriptRoot\Output\DashboardEasy04_00.html {
    Section -Name 'Test2' -Collapsable {
        Container {
            Panel {
                Table -DataTable $Process -PreContent { New-HTMLText -Text 'TestPostContent' -Color DodgerBlue -Alignment center } -Buttons @() -DisableSearch -DisablePaging -HideFooter
            }
            Panel {
                Table -DataTable $Process -PreContent { New-HTMLText -Text 'TestPostContent' -Color DodgerBlue -Alignment center } -Buttons @() -DisableSearch -DisablePaging -HideFooter
            }
        }
    }
}
```

I guess that's what you mean by elegant right? You should also remember that EmailTable is also there (alias to New-HTMLTable). 

So to summarize:
- EmailBody/Email will get the ability to just produce HTML without need to send email. The workaround now would be to use Email -FilePath $Path with -WhatIf parameter which would prevent sending email and at the same time create HTML content into file. 
- Dashimo style allows you to get similar stuff as Email

Unless you require something else?

**Comment by PrzemyslawKlys** [comment]
You will be able to use EmailBody or Email with -OutputHTML the way you want. Both options are supported.

Examples here: https://github.com/EvotecIT/PSWriteHTML/tree/master/Examples/Example38-Emails

**Comment by PrzemyslawKlys** [comment]
N
